# "Simple Neural Network"
> "Simple neural network to predict the function f(x) = 10*x"

- toc: false
- branch: master
- badges: true
- comments: true
- categories: [neural networks, keras, linear regression]
- image: images/some_folder/your_image.png
- hide: false
- search_exclude: true
- metadata_key1: metadata_value1
- metadata_key2: metadata_value2

### In this notebook, we will use a very simple neural network to make predictions. We will see if our simple neural network can predict the function f(x) = 10*x. 

#### This notebook is motivated from one of the [videos](http://www.youtube.com/watch?v=_Z9TRANg4c0) on Machine Learning Foundations by Google Developers.  

#### Let's start by importing some useful libraries


In [1]:
import tensorflow as tf
import numpy as np
from tensorflow import keras # We use keras API to define our model

#### Let's now create the simplest possible neural network i.e., with 1 layer and 1 neuron
> 

In [2]:
model = tf.keras.Sequential([keras.layers.Dense(units = 1, input_shape = [1])]) 

#### Let's compile our model by specifying an optimizer and loss function 


In [3]:

model.compile(optimizer = 'sgd', loss = 'mean_squared_error') # 'sgd' stands for stochastic gradient descent

In [4]:
# Time to provide the data to our NN

X = np.array([-1,0,1,2,3,4,5,6,7,8,9,10], dtype = float)
Y = np.array([-10,0,10,20,30,40,50,60,70,80,90,100], dtype = float)


#### In the above cell, you can already guess the function by looking at the x and y values. 

#### Let's train our neural network...

In [5]:
# Training the NN

model.fit(X,Y,epochs = 500)

Train on 12 samples
Epoch 1/500
12/12 [==============================] - 0s 28ms/sample - loss: 4246.3638
Epoch 2/500
12/12 [==============================] - 0s 201us/sample - loss: 503.1161
Epoch 3/500
12/12 [==============================] - 0s 214us/sample - loss: 60.3177
Epoch 4/500
12/12 [==============================] - 0s 172us/sample - loss: 7.9278
Epoch 5/500
12/12 [==============================] - 0s 157us/sample - loss: 1.7191
Epoch 6/500
12/12 [==============================] - 0s 156us/sample - loss: 0.9734
Epoch 7/500
12/12 [==============================] - 0s 176us/sample - loss: 0.8740
Epoch 8/500
12/12 [==============================] - 0s 153us/sample - loss: 0.8513
Epoch 9/500
12/12 [==============================] - 0s 200us/sample - loss: 0.8378
Epoch 10/500
12/12 [==============================] - 0s 176us/sample - loss: 0.8255
Epoch 11/500
12/12 [==============================] - 0s 155us/sample - loss: 0.8135
Epoch 12/500
12/12 [=============================

Epoch 98/500
12/12 [==============================] - 0s 130us/sample - loss: 0.2287
Epoch 99/500
12/12 [==============================] - 0s 180us/sample - loss: 0.2254
Epoch 100/500
12/12 [==============================] - 0s 184us/sample - loss: 0.2222
Epoch 101/500
12/12 [==============================] - 0s 182us/sample - loss: 0.2189
Epoch 102/500
12/12 [==============================] - 0s 128us/sample - loss: 0.2158
Epoch 103/500
12/12 [==============================] - 0s 265us/sample - loss: 0.2127
Epoch 104/500
12/12 [==============================] - 0s 191us/sample - loss: 0.2096
Epoch 105/500
12/12 [==============================] - 0s 227us/sample - loss: 0.2065
Epoch 106/500
12/12 [==============================] - 0s 197us/sample - loss: 0.2036
Epoch 107/500
12/12 [==============================] - 0s 131us/sample - loss: 0.2006
Epoch 108/500
12/12 [==============================] - 0s 235us/sample - loss: 0.1977
Epoch 109/500
12/12 [==============================] - 0

Epoch 194/500
12/12 [==============================] - 0s 145us/sample - loss: 0.0564
Epoch 195/500
12/12 [==============================] - 0s 131us/sample - loss: 0.0556
Epoch 196/500
12/12 [==============================] - 0s 136us/sample - loss: 0.0548
Epoch 197/500
12/12 [==============================] - 0s 141us/sample - loss: 0.0540
Epoch 198/500
12/12 [==============================] - 0s 147us/sample - loss: 0.0532
Epoch 199/500
12/12 [==============================] - 0s 110us/sample - loss: 0.0524
Epoch 200/500
12/12 [==============================] - 0s 147us/sample - loss: 0.0517
Epoch 201/500
12/12 [==============================] - 0s 142us/sample - loss: 0.0509
Epoch 202/500
12/12 [==============================] - 0s 176us/sample - loss: 0.0502
Epoch 203/500
12/12 [==============================] - 0s 115us/sample - loss: 0.0495
Epoch 204/500
12/12 [==============================] - 0s 135us/sample - loss: 0.0487
Epoch 205/500
12/12 [==============================] -

Epoch 290/500
12/12 [==============================] - 0s 119us/sample - loss: 0.0139
Epoch 291/500
12/12 [==============================] - 0s 117us/sample - loss: 0.0137
Epoch 292/500
12/12 [==============================] - 0s 159us/sample - loss: 0.0135
Epoch 293/500
12/12 [==============================] - 0s 140us/sample - loss: 0.0133
Epoch 294/500
12/12 [==============================] - 0s 121us/sample - loss: 0.0131
Epoch 295/500
12/12 [==============================] - 0s 125us/sample - loss: 0.0129
Epoch 296/500
12/12 [==============================] - 0s 135us/sample - loss: 0.0127
Epoch 297/500
12/12 [==============================] - 0s 192us/sample - loss: 0.0126
Epoch 298/500
12/12 [==============================] - 0s 199us/sample - loss: 0.0124
Epoch 299/500
12/12 [==============================] - 0s 175us/sample - loss: 0.0122
Epoch 300/500
12/12 [==============================] - 0s 152us/sample - loss: 0.0120
Epoch 301/500
12/12 [==============================] -

Epoch 386/500
12/12 [==============================] - 0s 138us/sample - loss: 0.0034
Epoch 387/500
12/12 [==============================] - 0s 163us/sample - loss: 0.0034
Epoch 388/500
12/12 [==============================] - 0s 192us/sample - loss: 0.0033
Epoch 389/500
12/12 [==============================] - 0s 222us/sample - loss: 0.0033
Epoch 390/500
12/12 [==============================] - 0s 209us/sample - loss: 0.0032
Epoch 391/500
12/12 [==============================] - 0s 208us/sample - loss: 0.0032
Epoch 392/500
12/12 [==============================] - 0s 152us/sample - loss: 0.0031
Epoch 393/500
12/12 [==============================] - 0s 146us/sample - loss: 0.0031
Epoch 394/500
12/12 [==============================] - 0s 167us/sample - loss: 0.0031
Epoch 395/500
12/12 [==============================] - 0s 124us/sample - loss: 0.0030
Epoch 396/500
12/12 [==============================] - 0s 219us/sample - loss: 0.0030
Epoch 397/500
12/12 [==============================] -

Epoch 481/500
12/12 [==============================] - 0s 188us/sample - loss: 8.5804e-04
Epoch 482/500
12/12 [==============================] - 0s 162us/sample - loss: 8.4565e-04
Epoch 483/500
12/12 [==============================] - 0s 138us/sample - loss: 8.3338e-04
Epoch 484/500
12/12 [==============================] - 0s 166us/sample - loss: 8.2129e-04
Epoch 485/500
12/12 [==============================] - 0s 166us/sample - loss: 8.0942e-04
Epoch 486/500
12/12 [==============================] - 0s 218us/sample - loss: 7.9771e-04
Epoch 487/500
12/12 [==============================] - 0s 116us/sample - loss: 7.8613e-04
Epoch 488/500
12/12 [==============================] - 0s 185us/sample - loss: 7.7477e-04
Epoch 489/500
12/12 [==============================] - 0s 205us/sample - loss: 7.6356e-04
Epoch 490/500
12/12 [==============================] - 0s 201us/sample - loss: 7.5247e-04
Epoch 491/500
12/12 [==============================] - 0s 174us/sample - loss: 7.4156e-04
Epoch 492/

#### One can see that as we started training our model, it made a very poor guess at the first epoch - loss value is huge. After few epochs, we see that the loss starts to go down and by the end of 500 epochs, loss is almost zero !

#### So now we have a model that is trained to learn the relationship between X and Y. Let's try to predict the value of X which the model have not seen before

In [6]:
print(model.predict([25.0]))

[[249.89447]]


#### We were expecting Y to be 250 but out NN gave us the value closeby to 250. This is because NN deals with probabilities. With only 12 data points, our model calculated that there is a very high probability that the relationship between X and Y is Y = 10X. As a result, the prediction for 25 is very close to 250 but not exactly 250. 

#### Feel free to write in the comments if you have any suggestions and comments. Thanks !